Should MatchFailureException capture an unreified tuple input?

We have tests like the following:

``` c#
        [Fact]
        public void UnmatchedInput_06()
        {
            var source =
@"using System;
public class C
{
    static void Main()
    {
        Console.WriteLine(M(1, 2));
        try
        {
            Console.WriteLine(M(1, 3));
        }
        catch (MatchFailureException ex)
        {
            Console.WriteLine($""{ex.GetType().Name}({ex.UnmatchedValue})"");
        }
    }
    public static int M(int x, int y) {
        return (x, y) switch { (1, 2) => 3 };
    }
}
namespace System
{
    public class MatchFailureException : InvalidOperationException
    {
        public MatchFailureException() {}
        public MatchFailureException(object unmatchedValue) => UnmatchedValue = unmatchedValue;
        public object UnmatchedValue { get; }
    }
}
";
            var compilation = CreatePatternCompilation(source);
            compilation.VerifyDiagnostics(
                // (17,23): warning CS8509: The switch expression does not handle all possible inputs (it is not exhaustive).
                //         return (x, y) switch { (1, 2) => 3 };
                Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustive, "switch").WithLocation(17, 23)
                );
            CompileAndVerify(compilation, expectedOutput: @"3
MatchFailureException()");
        }
```

Note that the exception throw in this case does not contain the `UnmatchedValue` that was input to the switch expression.  The reason the compiler does not pass that tuple to the exception constructor is that the compiler never constructs the tuple at all.  Rather the compiler produces the decision tree based on the individual values that were the input.  Because of that the generated code does not even have a runtime dependence on the presence of the `ValueTuple` library.

This is consistent with the [design of the `MatchFailureException` type](https://github.com/dotnet/corefx/issues/33284) and the LDM's specification that the captured value shall be provided when it is "trivial" to do so.  However, it would be possible for the compiler to produce code that constructs the tuple just for the failure case.

This is a tracking issue while we seek guidance on how to handle this situation.

--- a/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BasePatternSwitchLocalRewriter.cs
+++ b/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BasePatternSwitchLocalRewriter.cs
@@ -258,7 +258,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                 BoundDecisionDag decisionDag,
                 BoundExpression loweredSwitchGoverningExpression,
                 ArrayBuilder<BoundStatement> result,
-                out BoundExpression savedInputExpressionOpt)
+                out BoundExpression savedInputExpression)
             {
                 // Note that a when-clause can contain an assignment to a
                 // pattern variable declared in a different when-clause (e.g. in the same section, or
@@ -272,7 +272,7 @@ namespace Microsoft.CodeAnalysis.CSharp
 
                 if (canShareTemps)
                 {
-                    decisionDag = ShareTempsAndEvaluateInput(loweredSwitchGoverningExpression, decisionDag, expr => result.Add(_factory.ExpressionStatement(expr)), out savedInputExpressionOpt);
+                    decisionDag = ShareTempsAndEvaluateInput(loweredSwitchGoverningExpression, decisionDag, expr => result.Add(_factory.ExpressionStatement(expr)), out savedInputExpression);
                 }
                 else
                 {
@@ -280,7 +280,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                     BoundExpression inputTemp = _tempAllocator.GetTemp(InputTemp(loweredSwitchGoverningExpression));
                     Debug.Assert(inputTemp != loweredSwitchGoverningExpression);
                     result.Add(_factory.Assignment(inputTemp, loweredSwitchGoverningExpression));
-                    savedInputExpressionOpt = inputTemp;
+                    savedInputExpression = inputTemp;
                 }
 
                 // There is a hidden sequence point after evaluating the input at the start of the code to

--- a/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs
+++ b/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs
@@ -363,7 +363,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                 BoundExpression loweredInput,
                 BoundDecisionDag decisionDag,
                 Action<BoundExpression> addCode,
-                out BoundExpression savedInputExpressionOpt)
+                out BoundExpression savedInputExpression)
             {
                 var inputDagTemp = InputTemp(loweredInput);
                 if (loweredInput.Kind == BoundKind.Local || loweredInput.Kind == BoundKind.Parameter)
@@ -406,20 +406,20 @@ namespace Microsoft.CodeAnalysis.CSharp
                     // (though perhaps its component parts are used), then we can save the component parts
                     // and assign them into temps (or perhaps user variables) to avoid the creation of
                     // the tuple altogether.
-                    decisionDag = RewriteTupleInput(decisionDag, expr, addCode);
-                    savedInputExpressionOpt = null;
+                    decisionDag = RewriteTupleInput(decisionDag, expr, addCode, out savedInputExpression);
                 }
                 else
                 {
                     // Otherwise we emit an assignment of the input expression to a temporary variable.
                     BoundExpression inputTemp = _tempAllocator.GetTemp(inputDagTemp);
-                    savedInputExpressionOpt = inputTemp;
+                    savedInputExpression = inputTemp;
                     if (inputTemp != loweredInput)
                     {
                         addCode(_factory.AssignmentExpression(inputTemp, loweredInput));
                     }
                 }
 
+                Debug.Assert(savedInputExpression != null);
                 return decisionDag;
 
                 bool usesOriginalInput(BoundDecisionDagNode node)
@@ -450,11 +450,14 @@ namespace Microsoft.CodeAnalysis.CSharp
             /// values that have been stored in temps. This permits the caller to avoid creation of the tuple object
             /// itself. We also emit assignments of the tuple values into their corresponding temps.
             /// </summary>
+            /// <param name="savedInputExpression">An expression that produces the value of the original input if needed
+            /// by the caller.</param>
             /// <returns>A new decision dag that does not reference the input directly</returns>
             private BoundDecisionDag RewriteTupleInput(
                 BoundDecisionDag decisionDag,
                 BoundObjectCreationExpression loweredInput,
-                Action<BoundExpression> addCode)
+                Action<BoundExpression> addCode,
+                out BoundExpression savedInputExpression)
             {
                 int count = loweredInput.Arguments.Length;
                 var tupleElementEvaluated = new bool[count];
@@ -462,18 +465,26 @@ namespace Microsoft.CodeAnalysis.CSharp
 
                 // If any remaining input elements remain unevaluated, evaluate them now
                 var originalInput = BoundDagTemp.ForOriginalInput(loweredInput.Syntax, loweredInput.Type);
+                var newArguments = ArrayBuilder<BoundExpression>.GetInstance(loweredInput.Arguments.Length);
                 for (int i = 0; i < count; i++)
                 {
+                    var field = loweredInput.Type.TupleElements[i].CorrespondingTupleField;
+                    Debug.Assert(field != null);
+                    var expr = loweredInput.Arguments[i];
+                    var fieldFetchEvaluation = new BoundDagFieldEvaluation(expr.Syntax, field, originalInput);
+                    var temp = new BoundDagTemp(expr.Syntax, expr.Type, fieldFetchEvaluation, 0);
                     if (!tupleElementEvaluated[i])
                     {
-                        var expr = loweredInput.Arguments[i];
-                        var field = loweredInput.Type.TupleElements[i].CorrespondingTupleField;
-                        Debug.Assert(field != null);
-                        var fieldFetchEvaluation = new BoundDagFieldEvaluation(expr.Syntax, field, originalInput);
-                        var temp = new BoundDagTemp(expr.Syntax, expr.Type, fieldFetchEvaluation, 0);
                         storeToTemp(temp, expr);
                     }
+
+                    newArguments.Add(_tempAllocator.GetTemp(temp));
                 }
+
+                savedInputExpression = loweredInput.Update(
+                    loweredInput.Constructor, arguments: newArguments.ToImmutableAndFree(), loweredInput.ArgumentNamesOpt, loweredInput.ArgumentRefKindsOpt,
+                    loweredInput.Expanded, loweredInput.ArgsToParamsOpt, loweredInput.ConstantValueOpt,
+                    loweredInput.InitializerExpressionOpt, loweredInput.BinderOpt, loweredInput.Type);
 
                 return rewrittenDag;
 

--- a/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_SwitchExpression.cs
+++ b/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_SwitchExpression.cs
@@ -44,7 +44,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                 var outerVariables = ArrayBuilder<LocalSymbol>.GetInstance();
                 var loweredSwitchGoverningExpression = _localRewriter.VisitExpression(node.Expression);
                 BoundDecisionDag decisionDag = ShareTempsIfPossibleAndEvaluateInput(
-                    node.DecisionDag, loweredSwitchGoverningExpression, result, out BoundExpression optionalSavedInput);
+                    node.DecisionDag, loweredSwitchGoverningExpression, result, out BoundExpression savedInputExpression);
+                Debug.Assert(savedInputExpression != null);
 
                 // lower the decision dag.
                 (ImmutableArray<BoundStatement> loweredDag, ImmutableDictionary<SyntaxNode, ImmutableArray<BoundStatement>> switchSections) =
@@ -89,10 +90,9 @@ namespace Microsoft.CodeAnalysis.CSharp
                     result.Add(_factory.Label(node.DefaultLabel));
                     var objectType = _factory.SpecialType(SpecialType.System_Object);
                     var thrownExpression =
-                        (optionalSavedInput != null &&
-                                implicitConversionExists(optionalSavedInput, objectType) &&
+                        (implicitConversionExists(savedInputExpression, objectType) &&
                                 _factory.WellKnownMember(WellKnownMember.System_Runtime_CompilerServices_SwitchExpressionException__ctorObject, isOptional: true) is MethodSymbol exception1)
-                            ? _factory.New(exception1, _factory.Convert(objectType, optionalSavedInput)) :
+                            ? _factory.New(exception1, _factory.Convert(objectType, savedInputExpression)) :
                         (_factory.WellKnownMember(WellKnownMember.System_Runtime_CompilerServices_SwitchExpressionException__ctor, isOptional: true) is MethodSymbol exception0)
                             ? _factory.New(exception0) :
                         _factory.New(_factory.WellKnownMethod(WellKnownMember.System_InvalidOperationException__ctor));

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs
@@ -1998,7 +1998,7 @@ namespace System.Runtime.CompilerServices
 {
     public class SwitchExpressionException : InvalidOperationException
     {
-        public SwitchExpressionException() {}
+        public SwitchExpressionException() => throw null;
         public SwitchExpressionException(object unmatchedValue) => UnmatchedValue = unmatchedValue;
         public object UnmatchedValue { get; }
     }
@@ -2040,7 +2040,7 @@ namespace System.Runtime.CompilerServices
 {
     public class SwitchExpressionException : InvalidOperationException
     {
-        public SwitchExpressionException() {}
+        public SwitchExpressionException() => throw null;
         public SwitchExpressionException(object unmatchedValue) => UnmatchedValue = unmatchedValue;
         public object UnmatchedValue { get; }
     }
@@ -2088,7 +2088,7 @@ namespace System.Runtime.CompilerServices
     public class SwitchExpressionException : InvalidOperationException
     {
         public SwitchExpressionException() {}
-        public SwitchExpressionException(object unmatchedValue) => UnmatchedValue = unmatchedValue;
+        public SwitchExpressionException(object unmatchedValue) => throw null;
         public object UnmatchedValue { get; }
     }
 }
@@ -2945,7 +2945,7 @@ namespace System.Runtime.CompilerServices
 {
     public class SwitchExpressionException : InvalidOperationException
     {
-        public SwitchExpressionException() {}
+        public SwitchExpressionException() => throw null;
         public SwitchExpressionException(object unmatchedValue) => UnmatchedValue = unmatchedValue;
         public object UnmatchedValue { get; }
     }
@@ -3029,6 +3029,72 @@ A(5, 6).Deconstruct
 A(5, 6).Y
 A(2, 3).Y
 8");
+        }
+
+        [Fact]
+        public void MissingValueTuple()
+        {
+            var source = @"
+class Program
+{
+    static void Main()
+    {
+    }
+    int M(int x, int y)
+    {
+        return (x, y) switch { (1, 2) => 1, _ => 2 };
+    }
+}
+";
+            var compilation = CreateCompilationWithMscorlib40(source);
+            compilation.VerifyDiagnostics(
+                // (9,16): error CS8179: Predefined type 'System.ValueTuple`2' is not defined or imported
+                //         return (x, y) switch { (1, 2) => 1, _ => 2 };
+                Diagnostic(ErrorCode.ERR_PredefinedValueTupleTypeNotFound, "(x, y)").WithArguments("System.ValueTuple`2").WithLocation(9, 16)
+                );
+        }
+
+        [Fact]
+        public void UnmatchedInput_07()
+        {
+            var source =
+@"using System; using System.Runtime.CompilerServices;
+public class C
+{
+    static void Main()
+    {
+        Console.WriteLine(M(1, 2));
+        try
+        {
+            Console.WriteLine(M(1, 3));
+        }
+        catch (SwitchExpressionException ex)
+        {
+            Console.WriteLine($""{ex.GetType().Name}({ex.UnmatchedValue})"");
+        }
+    }
+    public static int M(int x, int y, int a = 3, int b = 4, int c = 5, int d = 6, int e = 7, int f = 8, int g = 9) {
+        return (x, y, a, b, c, d, e, f, g) switch { (1, 2, _, _, _, _, _, _, _) => 3 };
+    }
+}
+namespace System.Runtime.CompilerServices
+{
+    public class SwitchExpressionException : InvalidOperationException
+    {
+        public SwitchExpressionException() => throw null;
+        public SwitchExpressionException(object unmatchedValue) => UnmatchedValue = unmatchedValue;
+        public object UnmatchedValue { get; }
+    }
+}
+";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics(
+                // (17,44): warning CS8509: The switch expression does not handle all possible inputs (it is not exhaustive).
+                //         return (x, y, a, b, c, d, e, f, g) switch { (1, 2, _, _, _, _, _, _, _) => 3 };
+                Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustive, "switch").WithLocation(17, 44)
+                );
+            CompileAndVerify(compilation, expectedOutput: @"3
+SwitchExpressionException((1, 3, 3, 4, 5, 6, 7, 8, 9))");
         }
     }
 }

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs
@@ -2052,7 +2052,7 @@ namespace System.Runtime.CompilerServices
                 //             _ = (1, 2) switch { (3, 4) => 1 };
                 Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustive, "switch").WithLocation(8, 24)
                 );
-            CompileAndVerify(compilation, expectedOutput: "SwitchExpressionException()");
+            CompileAndVerify(compilation, expectedOutput: "SwitchExpressionException((1, 2))");
         }
 
         [Fact]
@@ -2958,7 +2958,7 @@ namespace System.Runtime.CompilerServices
                 Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustive, "switch").WithLocation(17, 23)
                 );
             CompileAndVerify(compilation, expectedOutput: @"3
-SwitchExpressionException()");
+SwitchExpressionException((1, 3))");
         }
 
         [Fact]

